0634-cpufreq-Add-Hardware-P-State-HWP-driver.patch

```diff
@@ -0,0 +1,731 @@
+         switch ( boot_cpu_data.x86_vendor )
+         {
+         case X86_VENDOR_INTEL:
+-            ret = cpufreq_register_driver(&acpi_cpufreq_driver);
++            if (hwp_available())
++                ret = hwp_register_driver();
++            else
++                ret = cpufreq_register_driver(&acpi_cpufreq_driver);
+             break;
+ 
+         case X86_VENDOR_AMD:
+diff --git a/xen/arch/x86/acpi/cpufreq/hwp.c b/xen/arch/x86/acpi/cpufreq/hwp.c
+new file mode 100644
+index 000000000000..f8e6fdbd416f
+--- /dev/null
++++ b/xen/arch/x86/acpi/cpufreq/hwp.c
+@@ -0,0 +1,533 @@
++/*
++ * hwp.c cpufreq driver to run Intel Hardware P-States (HWP)
++ *
++ * Copyright (C) 2021 Jason Andryuk <jandryuk@gmail.com>
++ *
++ * This program is free software; you can redistribute it and/or modify
++ * it under the terms of the GNU General Public License as published by
++ * the Free Software Foundation; either version 2 of the License, or (at
++ * your option) any later version.
++ *
++ * This program is distributed in the hope that it will be useful, but
++ * WITHOUT ANY WARRANTY; without even the implied warranty of
++ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
++ * General Public License for more details.
++ *
++ * You should have received a copy of the GNU General Public License along
++ * with this program; If not, see <http://www.gnu.org/licenses/>.
++ */
++
++#include <xen/cpumask.h>
++#include <xen/init.h>
++#include <xen/param.h>
++#include <xen/xmalloc.h>
++#include <asm/msr.h>
++#include <asm/io.h>
++#include <acpi/cpufreq/cpufreq.h>
++
++static bool feature_hwp;
++static bool feature_hwp_notification;
++static bool feature_hwp_activity_window;
++static bool feature_hwp_energy_perf;
++static bool feature_hwp_pkg_level_ctl;
++static bool feature_hwp_peci;
++
++static bool feature_hdc;
++static bool feature_fast_msr;
++
++bool opt_hwp = true;
++boolean_param("hwp", opt_hwp);
++
++union hwp_request
++{
++    struct
++    {
++        uint64_t min_perf:8;
++        uint64_t max_perf:8;
++        uint64_t desired:8;
++        uint64_t energy_perf:8;
++        uint64_t activity_window:10;
++        uint64_t package_control:1;
++        uint64_t reserved:16;
++        uint64_t activity_window_valid:1;
++        uint64_t energy_perf_valid:1;
++        uint64_t desired_valid:1;
++        uint64_t max_perf_valid:1;
++        uint64_t min_perf_valid:1;
++    };
++    uint64_t raw;
++};
++
++struct hwp_drv_data
++{
++    union
++    {
++        uint64_t hwp_caps;
++        struct
++        {
++            uint64_t hw_highest:8;
++            uint64_t hw_guaranteed:8;
++            uint64_t hw_most_efficient:8;
++            uint64_t hw_lowest:8;
++            uint64_t hw_reserved:32;
++        };
++    };
++    union hwp_request curr_req;
++    uint16_t activity_window;
++    uint8_t minimum;
++    uint8_t maximum;
++    uint8_t desired;
++    uint8_t energy_perf;
++};
```

Comment:
Not sure if padding is an issue, but if it is:

```suggestion
+    uint16_t padding;
+};
```

Reply:
@DemiMarie I don't think padding matters - hwp_drv_data is a xen-internal struct to track HWP capabilities and current settings.

@marmarek It looks like you picked up v1 of my patchset.  My v2 addressed some feedback, and dropped the "fast MSR" support.  There aren't too many changes though.  The 0634-cpufreq-Add-Hardware-P-State-HWP-driver.patch  v1 commit message says "it hardcodes the default 0x80 (out of 0x0-0xff) energy/performance preference.", but the v2 message correctly states "It will run with the default values, which should be the default 0x80 (out of 0x0-0xff) energy/performance preference."  That is, it just uses the current values the system booted with unless they are changed by the user.

https://lore.kernel.org/xen-devel/20220810192944.102135-1-jandryuk@gmail.com/#T

I am hoping to resubmit a v3 for upstream after I take care of some other pending work.